### PR TITLE
Feature/candidate invitaition

### PR DESCRIPTION
**Pull request by Sumit-Dhandale**
invitation link to candidate via mail

### CODE REVIEW

File: app/controllers/candidates_controller.rb [hunk 1]
```diff
@@ -1,4 +1,32 @@
...
+    candidate_emails.each do |candidate_email|
+      user = Candidate.new(email: candidate_email)
+      drive_candidate = DrivesCandidate.new(drive_id: params[:drive_id], candidate_id: user.id) if user.save
+
+      next unless user.present? && drive_candidate.save
+
+      drive_candidate.generate_token!
+      begin
+        CandidateMailer.invitation_email(user, drive_candidate).deliver_later
+      rescue StandardError => e
```
Comment (rahulknojha) [inline app/controllers/candidates_controller.rb:19]: what kind of exception will occur here ? Can you please elaborate ?

File: app/mailers/candidate_mailer.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class CandidateMailer < ApplicationMailer
+  def invitation_email(user, drive_candidate)
+    @token = drive_candidate.token.to_s
+    mail(to: user.email, from: ENV['MAIL_USERNAME'], subject: 'Invitation for coding round',
```
Comment (rahulknojha) [inline app/mailers/candidate_mailer.rb:6]: subject needs to be proper, should include organisation name as well

File: app/models/drives_candidate.rb [hunk 3]
```diff
@@ -1,5 +1,17 @@
 class DrivesCandidate < ApplicationRecord
   belongs_to :drive
   belongs_to :candidate
+
+  def generate_token!
+    self.token = generate_token
+    self.email_sent_at = Time.now.utc
+    save!
```
Comment (rahulknojha) [inline app/models/drives_candidate.rb:8]: what if it raises an exception ?

File: config/application.rb [hunk 4]
```diff
@@ -35,5 +35,13 @@ class Application < Rails::Application
     config.middleware.use ActionDispatch::Flash
     config.middleware.use ActionDispatch::Cookies
     config.middleware.use ActionDispatch::Session::CookieStore
+    config.before_configuration do
+      env_file = File.join(Rails.root, 'config', 'local_env.yml')
```
Comment (rahulknojha) [inline config/application.rb:39]: please update as discussed in other PR

File: spec/controllers/candidates_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe CandidatesController, type: :controller do
+  let(:user) { double Candidate, first_name: 'abc', email: 'abc@gmail.com' }
+  let(:drive_candidate) { double DrivesCandidate, token: 'abcdefgjijklmnop' }
+
+  it 'sends an email' do
+    expect { CandidateMailer.invitation_email(user, drive_candidate).deliver_now }.to change {
```
Comment (rahulknojha) [inline spec/controllers/candidates_controller_spec.rb:10]: you are writing test cases for controller but you have not called the controller action, and manually repeating the code here. I m not sure what you are trying to test here

**Review by rahulknojha — CHANGES_REQUESTED** [review]

File: app/models/drives_candidate.rb (hunk 3 above)
Comment (somnath77) [reply app/models/drives_candidate.rb]: handled exception

File: spec/controllers/candidates_controller_spec.rb (hunk 5 above)
Comment (somnath77) [reply spec/controllers/candidates_controller_spec.rb]: Added corrected unit tests

File: app/mailers/candidate_mailer.rb (hunk 2 above)
Comment (somnath77) [reply app/mailers/candidate_mailer.rb]: Added detailed subject

File: app/controllers/candidates_controller.rb (hunk 1 above)
Comment (somnath77) [reply app/controllers/candidates_controller.rb]: handled specific exceptions

File: app/models/drives_candidate.rb [hunk 6]
```diff
@@ -3,4 +3,21 @@
...
   belongs_to :candidate
+
+  def generate_token!
+    self.token = generate_token
+    self.email_sent_at = Time.now.utc
+    begin
+      save!
+    rescue ActiveRecord::RecordInvalid, ActiveRecord::RecordNotUnique => e
+      drives_candidate = exists?
+      raise e if drives_candidate.nil?
```
Comment (rahulknojha) [inline app/models/drives_candidate.rb:14]: how this can be nil ? what will be the scenario ?

**Review by rahulknojha — DISMISSED** [review]

**Review by rahulknojha — APPROVED** [review]

**Pull request merged**